crawl-ref/source/mon-place.cc

```diff
@@ -160,51 +156,78 @@ bool monster_habitable_grid(monster_type mt,
     // Its tentacles can, and will, though.
     if ((actual_grid == DNGN_SHALLOW_WATER
         || actual_grid == DNGN_TOXIC_BOG)
-        && mt == MONS_KRAKEN)
+        && avoid_shallow_water)
     {
         return false;
     }
 
-    const dungeon_feature_type feat_preferred =
-        habitat2grid(mons_class_primary_habitat(mt));
-    const dungeon_feature_type feat_nonpreferred =
-        habitat2grid(mons_class_secondary_habitat(mt));
-
     // If the caller insists on a specific feature type, try to honour
     // the request. This allows the builder to place amphibious
     // creatures only on land, or flying creatures only on lava, etc.
     if (wanted_grid != DNGN_UNSEEN
-        && monster_habitable_grid(mt, wanted_grid, DNGN_UNSEEN))
+        && _monster_habitable_grid(wanted_grid, DNGN_UNSEEN, monster_habitat,
+            is_flying, avoid_shallow_water, can_stand_on_malign_gateway))
     {
         return _feat_compatible(wanted_grid, actual_grid);
     }
 
     if (actual_grid == DNGN_MALIGN_GATEWAY)
-    {
-        if (mt == MONS_ELDRITCH_TENTACLE
-            || mt == MONS_ELDRITCH_TENTACLE_SEGMENT)
-        {
-            return true;
-        }
-        else
-            return false;
-    }
+        return can_stand_on_malign_gateway;
+
+    const dungeon_feature_type feat_preferred =
+        habitat2grid(primary_habitat(monster_habitat));
+    const dungeon_feature_type feat_nonpreferred =
+        habitat2grid(secondary_habitat(monster_habitat));
+
+    if (_feat_compatible(feat_preferred, actual_grid))
+        return true;
 
-    if (_feat_compatible(feat_preferred, actual_grid)
-        || _feat_compatible(feat_nonpreferred, actual_grid))
+    if (feat_preferred != feat_nonpreferred
+        && _feat_compatible(feat_nonpreferred, actual_grid))
     {
         return true;
     }
 
     // [dshaligram] Flying creatures are all HT_LAND, so we
     // only have to check for the additional valid grids of deep
     // water and lava.
-    if (_hab_requires_mon_flight(actual_grid) && (mons_class_flag(mt, M_FLIES)))
+    if (_hab_requires_mon_flight(actual_grid) && is_flying)
         return true;
 
     return false;
 }
 
+/**
+ * Can monsters of this class survive on actual_grid?
+ *
+ * @param mt the monster class to check against
+ * @param actual_grid the terrain feature being checked
+ * @param wanted_grid if == DNGN_UNSEEN, or if the monster can't survive on it,
+ *                    ignored. Otherwise, return false even if actual_grid is
+ *                    survivable, if actual_grid isn't similar to wanted_grid.
+ */
+bool monster_habitable_grid(monster_type mt,
```

Comment:
Personally, I like this new function a lot and feel like it'd be better to use universally than the function right after, since in the header file you even mention how it has the limitation of `// Can't handle krakens and eldritch tentacles as their habitats are special`.

That said, reading through your changes to `spl-summoning.cc`, I understand the want for the new function and like the concept of determining proper habitable tiles for to-be-summoned creatures. However, I think even the starting spell Summon Mammals shows a potential issue with it—specifically, the fact that flying bats can be summoned from the spell, but quokkas and rats do not fly, but would still be affected by the `flying=true` being used here.

```c++
spret cast_summon_small_mammal(int pow, god_type god, bool fail)
{
    if (!_can_summon_monster(you, HT_LAND, true))  // <-- Only one of the small mammals picks can actually fly
    {
        mpr("There is no available space!");
        return spret::abort;
    }
```

That said, I think this might be the best approach we could take as of now if we want to be able to generically find out if ANY of the possibly summoned creatures would be able to be summoned in open spaces near the caster.

Two things could occur as a result of these checks as I see it:
1. Mass deaths of any summoned HT_LAND creature with no flying on a drowning or lava cell
2. If other checks force only habitable creatures to be summoned, we could force nothing but bats and other flying creatures to be summoned from otherwise random spells

That said, I'm *very* new to the codebase and don't have expertise here, so feel free to ignore this comment if there are other checks or conditions that address those concerns already. 😄

Reply:
I'm honestly thinking of replacing the `_can_summon_monster` function with a function that takes an array of monster types and returns true if we can summon any of them seeing that's want `_can_summon_monster` is effectively being used for anyway. This probably wouldn't be as efficient, but it would mean that I wouldn't have to expose the hacks used to handle krakens and eldritch tentacles. Same with replacing the `monster_habitable_grid` function that takes a habitat.

Regarding casting summon small mammals with only space available for bats, I haven't changed the behavior here, you will have a chance of getting a bat and a chance of nothing happening. Although it would be nice to force only bats or prompt the player in this situation, that feels to me like a separate change to not letting the player waste a turn and magic points summoning when it would have no effect.

Reply:
Yeah, a uselessness check with a prompt *may* make sense, but prompts do feel (personally) so jarring it might be better for the player to just learn under what circumstances a spell makes sense to cast.

If we wanted to get real technical or feature heavy, I can imagine a HUD element and flavor text warning of the spell’s limited usefulness in certain circumstances.

Agreed in that this change is more singular in its purpose, so we can worry about those cases later.